client/src/session/expire.rs

```diff
@@ -34,40 +34,82 @@ pub async fn track_sessions_expiration(layer: SessionLayer) {
         // can last a few seconds especially in case of inactive sessions.
         let ping_futures = sessions
             .iter()
-            .map(|session| session.raw.keep_alive(expiration))
-            .collect::<Vec<_>>();
+            .map(|session| session.raw.keep_alive(expiration));
 
         let last_seen = futures::future::join_all(ping_futures).await;
 
         // Collect indices of Sessions to close.
-        let expired_idx = last_seen
-            .iter()
+        let (expired, live): (Vec<_>, Vec<_>) = last_seen
+            .into_iter()
             .enumerate()
-            .filter_map(|(i, timestamp)| match *timestamp + expiration < now {
-                true => Some(i),
-                false => None,
+            .partition(|(i, timestamp)| (*timestamp + expiration) < now);
+
+        log::trace!("Closing {} expired sessions.", expired.len());
+        close_sessions(layer.clone(), sessions, expired).await;
+
+        let node_info_futures = match layer.server_session().await {
```

Comment:
Calling `layer.server_session` here will try to initialize the session if it is not present. I'd rather just list the sessions and choose the one with default NodeId here so that there's no duplication of responsibilities with regartds to whemn a server session is being initialized (it might be already happening somewhere else)